File: compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/analysis/MustAnalysis.java
package org.jetbrains.kotlin.fir.resolve.dfa.analysis;

import java.util.*;

import org.jetbrains.kotlin.fir.resolve.dfa.input.*;

public class MustAnalysis {
	private Map<Instruction, AliasGraph> 	instrToGraph;
	private int 								idCounter;
	private Instruction 						lastInstr;

	public MustAnalysis() {
		instrToGraph = new HashMap<>();
		idCounter = 0;

		initMustAnalysis();
	}

	//Using to make linking of RVtoRVAT is possible
	private void initMustAnalysis() {
		Set<MustAnalysis> toInitInstruction = new HashSet<>();
		toInitInstruction.add(this);
		Instruction initInstruction = new PhiInstruction("-1", toInitInstruction);
		instrToGraph.put(initInstruction, new AliasGraph());
		lastInstr = initInstruction;
	}

	public MustAnalysis(MustAnalysis ma) {
		this();

		idCounter = ma.idCounter;
		int ch = 0;
		for (Instruction instr: ma.instrToGraph.keySet()) {
			ch++;
			instrToGraph.put(instr, ma.instrToGraph.get(instr));
		}
		lastInstr = ma.lastInstr;
	}

	public int getNewId() {
		idCounter++;
		return idCounter;
	}

	public AliasGraph getLastGraph() {
		return instrToGraph.get(lastInstr);
	}

	public void instrToFixpoint(LinkedHashMap<Instruction, Set<Instruction>> instrToPrev, LinkedHashMap<Instruction, Set<Instruction>> instrToNext) {
		Set<Instruction> reachableInstr = new TreeSet<>();
		for (Instruction instr : instrToPrev.keySet()) {
			for (Instruction instr1: instrToPrev.get(instr)) {
				reachableInstr.add(instr1);
				instrToGraph.put(instr1, new AliasGraph());
			}
		}
		for (Instruction instr : instrToNext.keySet()) {
			for (Instruction instr1: instrToNext.get(instr)) {
				reachableInstr.add(instr1);
				instrToGraph.put(instr1, new AliasGraph());
			}
		}
		applyToFixpoint(instrToPrev, instrToNext, reachableInstr);
	}

	// highly inefficient in many ways, and currently not really up to fixpoint
	//  - replaces the old graphs every time, does not just add deltas
	//  - does not optimize order of instructions to reach fixpoint
	private void applyToFixpoint(LinkedHashMap<Instruction, Set<Instruction>> instrToPrev, LinkedHashMap<Instruction, Set<Instruction>> instrToNext, Set<Instruction> reachableInstr) {
		boolean remLastInstr = true;
		Set<Instruction> instructionsToProcess = reachableInstr;

		for (int loop = 0; !instructionsToProcess.isEmpty() ; loop++) {
			Set<Instruction> deltaInstructions = new TreeSet<>();

			for (Instruction instr : instructionsToProcess) {
				if (instr == lastInstr) break;
				AliasGraph g = instrToGraph.get(instr);
				AliasGraph oldGraph = new AliasGraph(g);
				Set<Instruction> _nextInstructions = new TreeSet<>();

					Set<Instruction> predecessors = instrToPrev.get(instr);
					if (predecessors != null) {
						if (predecessors.size() == 1) {
							Iterator<Instruction> curAncestor = predecessors.iterator();
							Instruction ancestor = curAncestor.next();
							if (instr == ancestor) {
								if (!lastInstr.id.equals("-1") && remLastInstr) {
									remLastInstr = false;
									AliasGraph predecessorGraph = instrToGraph.get(lastInstr);
									g = new AliasGraph(predecessorGraph);
								} else {
									if (lastInstr.id.equals("-1")) {
										AliasGraph predecessorGraph = instrToGraph.get(lastInstr);
										g = new AliasGraph(predecessorGraph);
									}
								}
							} else {
								AliasGraph predecessorGraph = instrToGraph.get(ancestor);
								g = new AliasGraph(predecessorGraph);
							}
						}
						else if (predecessors.size() > 1) {
							Iterator<Instruction> curAncestor = predecessors.iterator();
							Instruction firstAncestor = curAncestor.next();
							Instruction secondAncestor = curAncestor.next();
							AliasGraph currentGraph = new AliasGraph();
							currentGraph.intersect(instrToGraph.get(firstAncestor), instrToGraph.get(secondAncestor));
							while (curAncestor.hasNext()) {
								AliasGraph newGraph = new AliasGraph();
								newGraph.intersect(currentGraph, instrToGraph.get(curAncestor.next()));
								currentGraph = newGraph;
							}
							g = currentGraph;
						}
					}

				boolean hasChange = applyInstruction(instr, g);

				if (!hasChange) {
					boolean areEqual = g.equals(oldGraph);
					hasChange = !areEqual;
				}

				if (hasChange) {
					Set<Instruction> nextSet = instrToNext.get(instr);
					if (nextSet != null)
						_nextInstructions.addAll(nextSet);

					deltaInstructions.addAll(_nextInstructions);
				}
				lastInstr = instr;
			}
			instructionsToProcess = deltaInstructions;
		}

			System.out.println("-- DONE --");
	}

	// Apply instruction semantics on the graph right before the instruction
	// and return whether there was any change
	boolean applyInstruction(Instruction csinstr, AliasGraph graph) {
		switch (csinstr.kind) {
			case MOVE:
				return handleMoveInstruction(csinstr, graph);
			case IDENTITY:
				return handleIdentityInstruction(csinstr, graph);
			case PHI:
				return handlePhiInstruction(csinstr, graph);
			default:
				//add info message
				return false;
		}
	}

	private boolean handleMoveInstruction(Instruction csinstr, AliasGraph graph) {
		instrToGraph.put(csinstr, graph);
		MoveInstruction instr = (MoveInstruction) csinstr;

		AliasNode right = graph.lookupVar(instr.fromVar);
		if (right == null) {
			right = new AliasNode();
			right.addVariable(instr.fromVar);
			graph.addNode(right);
		}
		AliasNode left = graph.lookupVar(instr.toVar);
		// from and to are already in the same node
		if (left == right) {
			return false;
		}

		if (left != null) {
			graph.removeVariableFromNode(left, instr.toVar);
		}
		graph.addVariableToNode(right, instr.toVar);

		graph.gcNodes();
		return true;
	}

	private boolean handleIdentityInstruction(Instruction csinstr, AliasGraph graph) {
		instrToGraph.put(csinstr, graph);
		IdentityInstruction instr = (IdentityInstruction) csinstr;

		AliasNode right = graph.lookupVar(instr.fromVar);
		if (right == null) {
			right = new AliasNode();
			right.addVariable(instr.fromVar);
			graph.addNode(right);
		}
		AliasNode left = graph.lookupVar(instr.toVar);
		// from and to are already in the same node
		if (left == right) {
			return false;
		}

		graph.addVariableToNode(right, instr.toVar);

		graph.gcNodes();
		return true;
	}

	private boolean handlePhiInstruction(Instruction i, AliasGraph graph) {
		instrToGraph.put(i, graph);
		PhiInstruction instr = (PhiInstruction) i;

		Iterator<MustAnalysis> itMust = instr.musts.iterator();
		MustAnalysis firstMust = itMust.next();
		if (firstMust == null) return false;
		AliasGraph currentGraph = new AliasGraph(graph);

		if (instr.musts.size() > 1) {
			MustAnalysis secondMust = itMust.next();
			currentGraph.intersect(firstMust.instrToGraph.get(firstMust.lastInstr), secondMust.instrToGraph.get(secondMust.lastInstr));
			while (itMust.hasNext()) {
				AliasGraph newGraph = new AliasGraph();
				newGraph.intersect(currentGraph, itMust.next().instrToGraph.get(itMust.next().lastInstr));
				currentGraph = newGraph;
			}
		} else {
			return false;
		}
		graph = new AliasGraph(currentGraph);

		graph.gcNodes();
		return true;
	}

	@Override
	public String toString() {
		String graphs = "";
		for (Instruction instruction: instrToGraph.keySet()) {
			graphs += instruction + ":\n";
			graphs += instrToGraph.get(instruction).toString();
		}
		return graphs;
	}
}

File: compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/analysis/MustAnalysis.kt
package org.jetbrains.kotlin.fir.resolve.dfa.analysis

import org.jetbrains.kotlin.fir.resolve.dfa.input.IdentityInstruction
import org.jetbrains.kotlin.fir.resolve.dfa.input.Instruction
import org.jetbrains.kotlin.fir.resolve.dfa.input.MoveInstruction
import org.jetbrains.kotlin.fir.resolve.dfa.input.PhiInstruction


class MustAnalysis() {
    private val instrToGraph: MutableMap<Instruction, AliasGraph>
    private var idCounter: Int
    private var lastInstr: Instruction
    val newId: Int
        get() {
            idCounter++
            return idCounter
        }
    val lastGraph: AliasGraph?
        get() = instrToGraph[lastInstr]

    init {
        instrToGraph = HashMap()
        idCounter = 0
        val toInitInstruction: MutableSet<MustAnalysis> = HashSet()
        toInitInstruction.add(this)
        val initInstruction: Instruction = PhiInstruction("-1", toInitInstruction)
        instrToGraph[initInstruction] = AliasGraph()
        lastInstr = initInstruction
    }

    constructor(ma: MustAnalysis) : this() {
        idCounter = ma.idCounter
        instrToGraph.clear()
        for (instr in ma.instrToGraph.keys) {
            instrToGraph[instr] = ma.instrToGraph[instr]!!
        }
        lastInstr = ma.lastInstr
    }

    fun instrToFixpoint(
        instrToPrev: LinkedHashMap<Instruction, Set<Instruction>>,
        instrToNext: LinkedHashMap<Instruction, Set<Instruction>>
    ) {
        val reachableInstr: MutableSet<Instruction> = LinkedHashSet()
        for (instr in instrToPrev.keys) {
            for (instr1 in instrToPrev[instr]!!) {
                reachableInstr.add(instr1)
                instrToGraph[instr1] = AliasGraph()
            }
        }
        for (instr in instrToNext.keys) {
            for (instr1 in instrToNext[instr]!!) {
                reachableInstr.add(instr1)
                instrToGraph[instr1] = AliasGraph()
            }
        }
        applyToFixpoint(instrToPrev, instrToNext, reachableInstr)
    }

    // highly inefficient in many ways, and currently not really up to fixpoint
    //  - replaces the old graphs every time, does not just add deltas
    //  - does not optimize order of instructions to reach fixpoint
    private fun applyToFixpoint(
        instrToPrev: LinkedHashMap<Instruction, Set<Instruction>>,
        instrToNext: LinkedHashMap<Instruction, Set<Instruction>>,
        reachableInstr: Set<Instruction>
    ) {
        var remLastInstr = true
        var instructionsToProcess = reachableInstr
        var loop = 0
        while (instructionsToProcess.isNotEmpty()) {
            val deltaInstructions: MutableSet<Instruction> = LinkedHashSet()
            for (instr in instructionsToProcess) {
                if (instr === lastInstr) break
                var g = instrToGraph[instr]
                val oldGraph = AliasGraph(g!!)
                val nextInstructions: MutableSet<Instruction> = LinkedHashSet()
                val predecessors = instrToPrev[instr]
                if (predecessors != null) {
                    if (predecessors.size == 1) {
                        val curAncestor = predecessors.iterator()
                        val ancestor = curAncestor.next()
                        if (instr === ancestor) {
                            if (lastInstr.id != "-1" && remLastInstr) {
                                remLastInstr = false
                                val predecessorGraph = instrToGraph[lastInstr]
                                g = AliasGraph(predecessorGraph!!)
                            } else {
                                if (lastInstr.id == "-1") {
                                    val predecessorGraph = instrToGraph[lastInstr]
                                    g = AliasGraph(predecessorGraph!!)
                                }
                            }
                        } else {
                            val predecessorGraph = instrToGraph[ancestor]
                            g = AliasGraph(predecessorGraph!!)
                        }
                    } else if (predecessors.size > 1) {
                        val curAncestor = predecessors.iterator()
                        val firstAncestor = curAncestor.next()
                        val secondAncestor = curAncestor.next()
                        var currentGraph = AliasGraph()
                        currentGraph.intersect(instrToGraph[firstAncestor], instrToGraph[secondAncestor])
                        while (curAncestor.hasNext()) {
                            val newGraph = AliasGraph()
                            newGraph.intersect(currentGraph, instrToGraph[curAncestor.next()])
                            currentGraph = newGraph
                        }
                        g = currentGraph
                    }
                }
                val info = applyInstruction(instr, g)
                g = info.second
                var hasChange = info.first
                if (!hasChange) {
                    val areEqual = g == oldGraph
                    hasChange = !areEqual
                }
                if (hasChange) {
                    val nextSet = instrToNext[instr]
                    if (nextSet != null) nextInstructions.addAll(nextSet)
                    deltaInstructions.addAll(nextInstructions)
                }
                lastInstr = instr
            }
            instructionsToProcess = deltaInstructions
            loop++
        }
        println("-- DONE --")
    }

    // Apply instruction semantics on the graph right before the instruction
    // and return whether there was any change
    fun applyInstruction(csinstr: Instruction, graph: AliasGraph): Pair<Boolean, AliasGraph> {
        return when (csinstr.kind) {
            Instruction.Kind.MOVE -> handleMoveInstruction(csinstr, graph)
            Instruction.Kind.IDENTITY -> handleIdentityInstruction(csinstr, graph)
            Instruction.Kind.PHI -> handlePhiInstruction(csinstr, graph)
        }
    }

    private fun handleMoveInstruction(csinstr: Instruction, graph: AliasGraph): Pair<Boolean, AliasGraph> {
        val hGraph = AliasGraph(graph)
        instrToGraph[csinstr] = hGraph
        val instr = csinstr as MoveInstruction
        var right = hGraph.lookupVar(instr.fromVar)
        if (right == null) {
            right = AliasNode()
            right.addVariable(instr.fromVar)
            hGraph.addNode(right)
        }
        val left = hGraph.lookupVar(instr.toVar)
        // from and to are already in the same node
        if (left === right) {
            return Pair(false, hGraph)
        }
        if (left != null) {
            hGraph.removeVariableFromNode(left, instr.toVar)
        }
        hGraph.addVariableToNode(right, instr.toVar)
        hGraph.gcNodes()
        return Pair(true, hGraph)
    }

    private fun handleIdentityInstruction(csinstr: Instruction, graph: AliasGraph): Pair<Boolean, AliasGraph> {
        val hGraph = AliasGraph(graph)
        instrToGraph[csinstr] = hGraph
        val instr = csinstr as IdentityInstruction
        var right = hGraph.lookupVar(instr.fromVar)
        if (right == null) {
            right = AliasNode()
            right.addVariable(instr.fromVar)
            hGraph.addNode(right)
        }
        val left = graph.lookupVar(instr.toVar)
        // from and to are already in the same node
        if (left === right) {
            return Pair(false, hGraph)
        }
        hGraph.addVariableToNode(right, instr.toVar)
        hGraph.gcNodes()
        return Pair(true, hGraph)
    }

    private fun handlePhiInstruction(csinstr: Instruction, graph: AliasGraph): Pair<Boolean, AliasGraph> {
        var hGraph = AliasGraph(graph)
        instrToGraph[csinstr] = hGraph
        val instr = csinstr as PhiInstruction
        val itMust: Iterator<MustAnalysis> = instr.musts.iterator()
        val firstMust = itMust.next()
        var currentGraph = AliasGraph(hGraph)
        if (instr.musts.size > 1) {
            val secondMust = itMust.next()
            currentGraph.intersect(firstMust.instrToGraph[firstMust.lastInstr], secondMust.instrToGraph[secondMust.lastInstr])
            while (itMust.hasNext()) {
                val newGraph = AliasGraph()
                newGraph.intersect(currentGraph, itMust.next().instrToGraph[itMust.next().lastInstr])
                currentGraph = newGraph
            }
        } else {
            return Pair(false, hGraph)
        }
        hGraph = AliasGraph(currentGraph)
        graph.gcNodes()
        return Pair(true, hGraph)
    }

    override fun toString(): String {
        var graphs = ""
        for (instruction in instrToGraph.keys) {
            graphs += """
                ${instruction.toString()}:
                
                """.trimIndent()
            graphs += instrToGraph[instruction].toString()
        }
        return graphs
    }
}